File: src/main/programs/array/EncodeAndDecodeString.java
package array;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class EncodeAndDecodeStrings {
    private static final char DELIMITER = '#';

    public String encode(List<String> strs) {
        StringBuilder encoded = new StringBuilder();

        for (String str : strs) {
            encoded.append(str.length())
                    .append(DELIMITER)
                    .append(str);
        }

        return encoded.toString();
    }

    public List<String> decode(String str) {
        List<String> decoded = new ArrayList<>();
        int i = 0;

        while (i < str.length()) {
            // Find the delimiter
            int delimiterIndex = str.indexOf(DELIMITER, i);

            // Get the length of next string
            int length = Integer.parseInt(str.substring(i, delimiterIndex));

            // Extract the string
            String substring = str.substring(delimiterIndex + 1,
                    delimiterIndex + 1 + length);
            decoded.add(substring);

            // Move pointer to start of next length
            i = delimiterIndex + 1 + length;
        }

        return decoded;
    }

    public static void main(String[] args) {
        EncodeAndDecodeStrings codec = new EncodeAndDecodeStrings();
        List<String> input = Arrays.asList("Hello", "World", "!");

        String encoded = codec.encode(input);
        List<String> decoded = codec.decode(encoded);

        System.out.println("Original: " + input);
        System.out.println("Encoded: " + encoded);
        System.out.println("Decoded: " + decoded);
    }
}

File: src/main/programs/array/EncodeAndDecodeString.kt
package array

class EncodeAndDecodeString {
    private val delimiter = '#'
    fun encodeString(string: List<String>): String {
        var stringBuilder = StringBuilder()
        for (str in string) {
            for (char in str) {
                stringBuilder.append(char.toInt())
                stringBuilder.append(CHAR_DELIMITER)
            }
            stringBuilder.append(STRING_DELIMITER)
        }
        return stringBuilder.toString()
    }

    fun encode(strs: List<String>): String =
        strs.joinToString("") { "${it.length}$delimiter$it" }

    fun decode(str: String): List<String> {
        val result = mutableListOf<String>()
        var i = 0

        while (i < str.length) {
            // Find delimiter
            val delimiterIndex = str.indexOf(delimiter, i)

            // Parse length
            val length = str.substring(i, delimiterIndex).toInt()

            // Extract string
            val substring = str.substring(delimiterIndex + 1,
                delimiterIndex + 1 + length)
            result.add(substring)

            // Move pointer
            i = delimiterIndex + 1 + length
        }

        return result
    }

   /* fun decodeString(string: String): List<String> {
        var stringBuilder=StringBuilder()
        var resultList= mutableListOf<String>()
        var i=0
        while (i in string.indices){

            while (string[i] != STRING_DELIMITER){
                var charIntegerValue=""
                while (string[i] != CHAR_DELIMITER) {
                    charIntegerValue += string[i]
                    i++
                }
                stringBuilder.append(charIntegerValue.toInt().toChar())
                i++
            }
            resultList.kotlin_fund_prog.add(stringBuilder.toString())
            stringBuilder.clear()
            i++
        }
        resultList
    }*/



    companion object{
        private const val CHAR_DELIMITER="|"
        private const val STRING_DELIMITER="/"
    }

}
fun main(){
    val codec = EncodeAndDecodeString()
    val input = listOf("Hello", "World", "!")

    val encoded = codec.encode(input)
    val decoded = codec.decode(encoded)

    println("Original: $input")
    println("Encoded: $encoded")
    println("Decoded: $decoded")
}

